### PR TITLE
Add many tests

### PR DESCRIPTION
**Pull request by jverbraeken**
- Relevant Issues: -
- Related Pull Requests: -

* * *

## What
This Pull Request adds to the repository many tests for many objects

- NoBluetoothException
- SerializationException
- CannotHappenException
- ServiceLocatorNoNetwork
- ATapData
- NoBluetoothException
- MultiTap
- DatabaseException
- NewDataReceivedEvent

## Why
This Pull Request is needed because we want to verify that everything works as intended

## How
This feature can be viewed/tested within the project by running the tests

## Alternative implementation
Other implementations that I've have considered are

### CODE REVIEW

File: app/src/main/java/com/nervousfish/nervousfish/exceptions/NoBluetoothException.java [hunk 1]
```diff
@@ -28,6 +21,16 @@ public NoBluetoothException(final String msg) {
     }
 
     /**
+     * Constructs a new NoBluetoothException that's thrown when there is an issue with
+     * deserialization.
+     *
+     * @param throwable The exception that occurred the layer above
+     */
+    public NoBluetoothException(final Throwable throwable) {
```
Comment (TheBonheurs) [inline app/src/main/java/com/nervousfish/nervousfish/exceptions/NoBluetoothException.java]: You sure the javadoc is correct on this one?

File: app/src/main/java/com/nervousfish/nervousfish/exceptions/SerializationException.java [hunk 2]
```diff
@@ -22,7 +22,7 @@ public SerializationException(final String msg) {
     }
 
     /**
-     * Constructs a new DeserializationException that's thrown when there is an issue with
+     * Constructs a new SerializationException that's thrown when there is an issue with
      * deserialization.
```
Comment (TheBonheurs) [inline app/src/main/java/com/nervousfish/nervousfish/exceptions/SerializationException.java]: Shouldn't it be Serialization, hence the name?

File: app/src/main/java/com/nervousfish/nervousfish/data_objects/VerificationMethod.java [hunk 3]
```diff
@@ -19,6 +19,9 @@
      * @param verificationMethodEnum The verification method this object should represent.
      */
     public VerificationMethod(final VerificationMethodEnum verificationMethodEnum) {
+        if (verificationMethodEnum == null) {
+            throw new IllegalArgumentException("The verification method may not be null");
+        }
         this.verificationMethodEnum = verificationMethodEnum;
```
Comment (ericcornelissen) [inline app/src/main/java/com/nervousfish/nervousfish/data_objects/VerificationMethod.java]: You didn't touch it, but can you rename this attribute to `verificationMethod` 😇

File: app/src/main/java/com/nervousfish/nervousfish/data_objects/tap/ATapData.java [hunk 4]
```diff
@@ -20,9 +22,12 @@ public ATapData() {
     /**
      * <b>Should only be used for testing purposes!</b>
      *
-     * @param timestamp The timestamp of the tap data
+     * @param timestamp The timestamp of the tap data (NonNull)
```
Comment (ericcornelissen) [inline app/src/main/java/com/nervousfish/nervousfish/data_objects/tap/ATapData.java]: Can you remove this and add: `@throws IllegalArgumentException If the provided {@code timestamp} is {@code null}`

File: app/src/main/java/com/nervousfish/nervousfish/modules/database/DatabaseException.java [hunk 5]
```diff
@@ -25,4 +29,30 @@ public DatabaseException(final Throwable throwable) {
         super(throwable);
     }
 
+    /**
+     * Deserialize the instance using readObject to ensure invariants and security.
+     *
+     * @param stream The serialized object to be deserialized
+     */
+    private void readObject(final ObjectInputStream stream) throws IOException, ClassNotFoundException {
```
Comment (ericcornelissen) [inline app/src/main/java/com/nervousfish/nervousfish/modules/database/DatabaseException.java]: These seem to be automatically generated? If so, can you remove them 😇

File: app/src/test/java/com/nervousfish/nervousfish/data_objects/tap/ATapDataTest.java [hunk 6]
```diff
@@ -14,6 +16,19 @@
 
     private static class Tmp extends ATapData {
```
Comment (ericcornelissen) [inline app/src/test/java/com/nervousfish/nervousfish/data_objects/tap/ATapDataTest.java]: Can you rename this private static class to `TestTapData` 😇

File: app/src/test/java/com/nervousfish/nervousfish/modules/constants/ConstantsTest.java [hunk 7]
```diff
@@ -0,0 +1,40 @@
...
+import org.junit.Before;
+import org.junit.Test;
+
+import static org.junit.Assert.*;
+import static org.mockito.Mockito.mock;
+
+/**
+ * ----------- NOTE ------------
+ * We cannot test much more than just that the results are not null.
+ * If we change the names of the
```
Comment (TheBonheurs) [inline app/src/test/java/com/nervousfish/nervousfish/modules/constants/ConstantsTest.java]: of the ... what?

Comment (ericcornelissen) [inline app/src/test/java/com/nervousfish/nervousfish/modules/constants/ConstantsTest.java]: I think you should use JavaDoc comments (`/** */`) for "notes"... Can you remove the `----------- NOTE ------------` and replace the `/** */` with normal comments (`//`)?

File: fooaccountInformation.json [hunk 8]
```diff
@@ -0,0 +1 @@
+[]
```
Comment (TheBonheurs) [inline fooaccountInformation.json]: What are these?

File: app/src/test/java/com/nervousfish/nervousfish/util/CircularListTest.java [hunk 9]
```diff
@@ -81,6 +81,7 @@ public void testEqualsHash() {
         list1.add(12);
         assertTrue(list3.equals(list4));
         assertTrue(list3.hashCode() == list4.hashCode());
+        assertFalse(list3.equals(null));
```
Comment (TheBonheurs) [inline app/src/test/java/com/nervousfish/nervousfish/util/CircularListTest.java]: AssertNotNull?

File: app/src/test/java/com/nervousfish/nervousfish/modules/pairing/events/BluetoothConnectedEventTest.java [hunk 10]
```diff
@@ -1,10 +1,22 @@
...
+import static org.mockito.Mockito.mock;
+
 public class BluetoothConnectedEventTest {
     @Test
     public void testBluetoothConnectedEvent() {
         new BluetoothConnectedEvent(null);
     }
+
+    @Test
+    public void testBluetoothConnectedEventGetThreadNull() {
```
Comment (ericcornelissen) [inline app/src/test/java/com/nervousfish/nervousfish/modules/pairing/events/BluetoothConnectedEventTest.java]: Not sure how this particular test (or even this whole test class for that matter) is useful 🤔 It doesn't seem to be testing anything useful...

File: foocontacts.json [hunk 11]
```diff
@@ -0,0 +1 @@
+[]
```
Comment (TheBonheurs) [inline foocontacts.json]: What are these?

File: app/src/test/java/com/nervousfish/nervousfish/service_locator/ServiceLocatorNoNetworkTest.java [hunk 12]
```diff
@@ -0,0 +1,154 @@
...
+import java.io.IOException;
+import java.io.ObjectInputStream;
+import java.io.ObjectOutputStream;
+
+import static org.junit.Assert.assertEquals;
+import static org.junit.Assert.assertNotNull;
+import static org.junit.Assert.assertTrue;
+import static org.mockito.Mockito.mock;
+
+public final class ServiceLocatorNoNetworkTest {
```
Comment (ericcornelissen) [inline app/src/test/java/com/nervousfish/nervousfish/service_locator/ServiceLocatorNoNetworkTest.java]: Shouldn't `ServiceLocatorNoNetwork` a test class itself, i.e. not part of the source code?

File: app/src/test/java/com/nervousfish/nervousfish/util/CircularListTest.java (hunk 9 above)
Comment (ericcornelissen) [inline app/src/test/java/com/nervousfish/nervousfish/util/CircularListTest.java]: You're already asserting that `list3` is equal to `list4`, and `list4` cannot be null here

File: fooaccountInformation.json (hunk 8 above)
Comment (ericcornelissen) [inline fooaccountInformation.json]: Can you remove this file (and `foocontacts.json`) from the PR 😇

File: app/src/main/java/com/nervousfish/nervousfish/exceptions/NoBluetoothException.java (hunk 1 above)
Comment (jverbraeken) [inline app/src/main/java/com/nervousfish/nervousfish/exceptions/NoBluetoothException.java]: Not completely sure

File: app/src/main/java/com/nervousfish/nervousfish/modules/database/DatabaseException.java (hunk 5 above)
Comment (jverbraeken) [inline app/src/main/java/com/nervousfish/nervousfish/modules/database/DatabaseException.java]: Hmm, I didn't generate it, talked with somebody from Forze about these kind of issues and he considered it good practice to explicitly declare objects that may be redundant at the time being but are not redundant anymore when class gets more functionality. When you add an attribute to a class you're much more likely to forget to add that not-so-obvious code that's needed as well. In this case I can imagine that when we add an attribute that describes the database file in which the error occurred, people will forget to add a Serialization Proxy to check if the database file is not initialized to null

File: app/src/test/java/com/nervousfish/nervousfish/util/CircularListTest.java (hunk 9 above)
Comment (jverbraeken) [inline app/src/test/java/com/nervousfish/nervousfish/util/CircularListTest.java]: Yeah, but there may be a bug in the equals method causing the result to be true when the input is null. We have to test this branch for full test coverage

File: app/src/main/java/com/nervousfish/nervousfish/exceptions/CannotHappenException.java [hunk 13]
```diff
@@ -12,7 +12,7 @@
     private static final long serialVersionUID = -7030163248191198652L;
 
     /**
-     * Constructor for a new NoBluetoothException with a message.
+     * Constructor for a new CannotHappenException with a message.
```
Comment (clenrock) [inline app/src/main/java/com/nervousfish/nervousfish/exceptions/CannotHappenException.java]: Is this actually used? Or only in the tests

Comment (jverbraeken) [inline app/src/main/java/com/nervousfish/nervousfish/exceptions/CannotHappenException.java]: Yes, it is used. For example when the machine learning considers an interval neither short nor long

File: app/src/main/java/com/nervousfish/nervousfish/modules/database/DatabaseException.java (hunk 5 above)
Comment (ericcornelissen) [inline app/src/main/java/com/nervousfish/nervousfish/modules/database/DatabaseException.java]: I think this is in conflict with your approach to not having to explicitly specify attributes (e.g. explicitly say `boolean myBoolean = false`). But sure, I'll let it go

File: app/src/main/java/com/nervousfish/nervousfish/exceptions/CannotHappenException.java (hunk 13 above)
Comment (ericcornelissen) [inline app/src/main/java/com/nervousfish/nervousfish/exceptions/CannotHappenException.java]: If it is only used then, shouldn't you rename it?

File: app/src/main/java/com/nervousfish/nervousfish/activities/RhythmCreateActivity.java [hunk 14]
```diff
@@ -296,7 +296,7 @@ private void addClosestTimestampToCluster(final long centerMean1, final long cen
                 this.clusterCenter2.add(closestPoint.getRight());
             } else {
                 LOGGER.error("A timestamp does neither belong to the short or long timestamp");
-                throw new CannotHappenException("This cannot happen");
+                throw new UnknownIntervalException("This cannot happen");
```
Comment (ericcornelissen) [inline app/src/main/java/com/nervousfish/nervousfish/activities/RhythmCreateActivity.java]: Can you edit the description as well 😇

**Comment by ericcornelissen** [comment]
Merging regardless of Travis CI status, unexpected failure of androidTest (like due to an emulator crash) 😕

**Pull request merged**